Libraries/PeriphDrivers/Source/SPIMSS/spimss_reva.c

```diff
@@ -231,6 +242,197 @@ int MXC_SPIMSS_RevA_MasterTrans(mxc_spimss_reva_regs_t *spi, spimss_reva_req_t *
+static mxc_dma_adv_config_t dma_adv_config_rx;
+static mxc_dma_srcdst_t srcdst_config_tx;
+static mxc_dma_srcdst_t srcdst_config_rx;
+/* ************************************************************************** */
+int MXC_SPIMSS_RevA_MasterTransDMA(mxc_spimss_reva_regs_t *spi, spimss_reva_req_t *req)
+{
+    // Calculating the transaction size in byte.
+    int sent_byte_len = ((req->len * (req->bits / 8)) - req->tx_num);
+    int tx_channel_id = -1;
+    int rx_channel_id = -1;
+    int spi_num = MXC_SPIMSS_GET_IDX((mxc_spimss_regs_t *)spi);
+
+    //Reset static variables.
+    // tx_channel_reg = NULL;
+    // rx_channel_reg = NULL;
+    memset(&dma_config_tx, 0, sizeof(dma_config_tx));
+    memset(&dma_config_rx, 0, sizeof(dma_config_rx));
+    memset(&srcdst_config_tx, 0, sizeof(srcdst_config_tx));
+    memset(&srcdst_config_rx, 0, sizeof(srcdst_config_rx));
+    memset(&dma_adv_config_tx, 0, sizeof(dma_adv_config_tx));
+    memset(&dma_adv_config_rx, 0, sizeof(dma_adv_config_rx));
+
+    // We use SPIMSS master trans function to send data from
+    if (E_NO_ERROR != MXC_SPIMSS_RevA_TransSetup(spi, req, 1)) {
+        return -1;
+    }
+
+    //Setting dmas fifo levels for tx and rx in spimss.
+    spi->dma &= ~(7 << MXC_F_SPIMSS_DMA_RX_FIFO_LVL_POS);
+    spi->dma &= ~(7 << MXC_F_SPIMSS_DMA_TX_FIFO_LVL_POS);
+
+    //Initialization of DMA
+    MXC_DMA_Init();
+
+    if ((true == states[spi_num].auto_dma_handlers) && (0 > states[spi_num].channelTx)) {
+        // Getting the available tx channel id.
+        tx_channel_id = MXC_DMA_AcquireChannel();
+        if (tx_channel_id < 0) {
+            return -1;
+        }
+        MXC_SPIMSS_RevA_SetTXDMAChannel(spi, tx_channel_id);
+
+        //Enable NVIC interrupt of the dma channel which will be used as TX channel.
+        //NVIC_EnableIRQ(MXC_DMA_CH_GET_IRQ(tx_channel_id));
+        MXC_NVIC_SetVector(MXC_DMA_CH_GET_IRQ(tx_channel_id), MXC_DMA_Handler);
+    } else {
+        tx_channel_id = MXC_SPIMSS_RevA_GetTXDMAChannel(spi);
+        if (tx_channel_id < 0) {
+            return E_BAD_STATE;
+        }
+    }
+
+    if ((true == states[spi_num].auto_dma_handlers) && (0 > states[spi_num].channelRx)) {
+        rx_channel_id = MXC_DMA_AcquireChannel();
+        if (rx_channel_id < 0) {
+            return -1;
+        }
+        MXC_SPIMSS_RevA_SetRXDMAChannel(spi, rx_channel_id);
+
+        //Enable NVIC interrupt of the dma channel which will be used as RX channel.
+        //NVIC_EnableIRQ(MXC_DMA_CH_GET_IRQ(rx_channel_id));
+        MXC_NVIC_SetVector(MXC_DMA_CH_GET_IRQ(rx_channel_id), MXC_DMA_Handler);
+    } else {
+        rx_channel_id = MXC_SPIMSS_RevA_GetRXDMAChannel(spi);
+    }
+
+    states[spi_num].req = req;
+
+    // TX dma configuration settings.
+    dma_config_tx.ch = tx_channel_id;
+    dma_config_tx.reqsel = MXC_DMA_REQUEST_SPI1TX;
+    dma_config_tx.dstwd = MXC_DMA_WIDTH_HALFWORD;
+    dma_config_tx.srcwd = MXC_DMA_WIDTH_HALFWORD;
```

Comment:
I think these DMA settings make sense, but this has the effect of enforcing 16-bit data alignment.  However, the request struct declares tx and rx data as `void *`...  

It seems like we should change the types in `spimss_reva_req` to enforce `uint16_t`...  but I don't have much experience with the SPIMSS block.  Maybe there is a good reason these are declared as void pointers?

Do you have the ability to test with characters less than 16-bits?  We may need to handle the left/right justification

See section 16.5 of the [UG](https://www.analog.com/media/en/technical-documentation/user-guides/max32660-user-guide.pdf).

![image](https://github.com/Analog-Devices-MSDK/msdk/assets/38844790/97ad59ed-a190-4863-aa22-f4e5e9309682)

Reply:
Hi @Jake-Carter , Thank you for your feedback. It was very helpful. I added this "spi->mode |= MXC_F_SPIMSS_REVA_MODE_TX_LJ;" line in MXC_SPIMSS_RevA_TransSetup function. So, if the bit size less than 8, transmit data is written to SPIMSS_DATA[7:0] and hardware shifts these bits to SPIMSS_DATA[15:8]. According to my tests, received data is as expected when we set this bit as indicated in the user guide for character transmission with less than 16 bits.